group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -822,21 +834,60 @@ public CompletableFuture<DeleteGroupsResponseData.DeletableGroupResultCollection
         });
 
         groupsByTopicPartition.forEach((topicPartition, groupList) -> {
-            CompletableFuture<DeleteGroupsResponseData.DeletableGroupResultCollection> future =
-                runtime.scheduleWriteOperation(
-                    "delete-groups",
-                    topicPartition,
-                    Duration.ofMillis(config.offsetCommitTimeoutMs()),
-                    coordinator -> coordinator.deleteGroups(context, groupList)
-                ).exceptionally(exception -> handleOperationException(
-                    "delete-groups",
-                    groupList,
-                    exception,
-                    (error, __) -> DeleteGroupsRequest.getErrorResultCollection(groupList, error),
-                    log
-                ));
-
-            futures.add(future);
+            CompletableFuture<DeleteGroupsResponseData.DeletableGroupResultCollection> shareFuture = deleteShareGroups(topicPartition, groupList)
```

Comment:
I really don't like that fact that we will obliterate the share-group state before checking whether the group is actually empty. This seems like a fundamental flaw to arranging the code like this.

Can the persister being invoked from the `GroupCoordinatorShard` instead?

Reply:
It was discouraged - we can do it ...

So, first we will send a list of all groups to GroupCoordinatorShard which will return a list of delete requests only for VALID SHARE GROUPS. Other groups will be ignored. This is the only extra work - we are only reading state from the a hashmap.

Reply:
This is possible but it would require to make the persister work with the event based workflows. Basically, we can not wait on the future from the shard and there is no way at the moment to react to it neither. We could make it work by wrapping the persister into a class which understand how to interface with the runtime/shard but it is more complicated than doing it in the service layer.

Reply:
@dajac Thanks for the response. We have a few other interactions which are a similar shape, so this is helpful guidance for those too.